lib/Dialect/MSFT/MSFTPasses.cpp

```diff
@@ -1670,6 +1670,113 @@ std::unique_ptr<Pass> createWireCleanupPass() {
 } // namespace msft
 } // namespace circt
 
+//===----------------------------------------------------------------------===//
+// Lower MSFT constructs
+//===----------------------------------------------------------------------===//
+
+namespace {
+
+struct LowerConstructsPass : public LowerConstructsBase<LowerConstructsPass>,
+                             PassCommon {
+  void runOnOperation() override;
+};
+} // anonymous namespace
+
+namespace {
+/// Lower MSFT's OutputOp to HW's.
+struct SystolicArrayOpLowering : public OpConversionPattern<SystolicArrayOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(SystolicArrayOp array, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+    MLIRContext *ctxt = getContext();
+    Location loc = array.getLoc();
+    Block &peBlock = array.pe().front();
+    rewriter.setInsertionPointAfter(array);
+
+    hw::ArrayType rowInputs =
+        hw::type_cast<hw::ArrayType>(array.rowInputs().getType());
+    IntegerType rowIdxType = rewriter.getIntegerType(
+        std::max(1u, llvm::Log2_64_Ceil(rowInputs.getSize())));
+    SmallVector<Value> rowValues;
+    for (size_t rowNum = 0, numRows = rowInputs.getSize(); rowNum < numRows;
+         ++rowNum) {
+      Value rowNumVal =
+          rewriter.create<hw::ConstantOp>(loc, rowIdxType, rowNum);
+      auto rowValue =
+          rewriter.create<hw::ArrayGetOp>(loc, array.rowInputs(), rowNumVal);
+      rowValue->setAttr("sv.namehint",
+                        StringAttr::get(ctxt, "row_" + Twine(rowNum)));
+      rowValues.push_back(rowValue);
+    }
+
+    hw::ArrayType colInputs =
+        hw::type_cast<hw::ArrayType>(array.colInputs().getType());
+    IntegerType colIdxType = rewriter.getIntegerType(
+        std::max(1u, llvm::Log2_64_Ceil(colInputs.getSize())));
+    SmallVector<Value> colValues;
+    for (size_t colNum = 0, numCols = colInputs.getSize(); colNum < numCols;
+         ++colNum) {
+      Value colNumVal =
+          rewriter.create<hw::ConstantOp>(loc, colIdxType, colNum);
+      auto colValue =
+          rewriter.create<hw::ArrayGetOp>(loc, array.colInputs(), colNumVal);
+      colValue->setAttr("sv.namehint",
+                        StringAttr::get(ctxt, "col_" + Twine(colNum)));
+      colValues.push_back(colValue);
+    }
+
+    SmallVector<Value> peOutputs;
+    for (Value rowValue : rowValues) {
```

Comment:
Is there any desire to generate SV for loops here? I don't know how people actually feel about `sv.for`, but it was somewhat recently brought up in another context: https://github.com/llvm/circt/issues/2759

Reply:
Yeah, I thought about that. In a subsequent PR, however, we'll have a non-uniform connectivity matrix -- producing an unrolled version of the systolic array will make generating that connectivity easier. There is an argument, however, for hiding the ugliness in a separate "connectivity" module that the `sv.for` loop connects to.